### ISSUE TITLE
Heavy stuttering on newer iOS devices

### ISSUE THREAD

**Issue by pip-byte** [godotengine/godot#38187]
**Godot version:** 3.2.1 stable

**OS/device including version:** iOS, MacOS

**Issue description:**
Godot runs buttery smooth on my iPhone 5s and iPhone X but stutters heavily on my 2020 iPad Pro and on my iMac. I tried..

- switching VSync on/off
- manually setting the physics refresh rate
- different physicsjitterfix values
- installing the smoothing-addon
- forcing 60fps in the preferences
- tried  _process and _physics_process

..but to no avail. The stuttering is quite severe and very distracting especially on a 12,9" iPad screen. It might have something to do with the Pro Motion Display of the iPad Pro witch supports variable frame rates up to 120Hz. The stuttering disappears from time to time and the game runs smooth for a few seconds but it is present most of the time.

We would like to publish our game but we can't do it with the heavy stuttering on newer devices.
We would probably get a lot of negative reviews because of the performance.

**Minimal reproduction project:**
Simple scene with any moving object (3D).

`VELOCITY = move_and_collide(Vector3(SPEED,0,0)*delta)`

**Comment by Calinou** [comment]
Try installing @lawnjelly's [smoothing-addon](https://github.com/lawnjelly/smoothing-addon) then configure it for your physics bodies. It should greatly help reduce stuttering.

Alternatively, increase the Physics Fps to 120, but keep in mind this will use much more CPU.

**Comment by Calinou** [comment]
@pip-byte Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

**Comment by Calinou** [comment]
Does Godot support rendering at above 60 FPS on iOS devices?

**Comment by pip-byte** [comment]
It stays at a solid 60FPS on all devices. I do not think this is a performance issue. The game works great on my very old iPhone 5s from 2013 and my 2017 iPhone X but stutters on my much more powerful 2020 iPad Pro. It also stutters on my iMac.

This has nothing to do with our game. It can be easily reproduced using a simple scene with a cube and a kinematic body.

**Comment by lawnjelly** [comment]
I've never used a variable frame rate display, but hopefully the following advice (which is also useful for fixed frame rate displays) will help, it sounds like you may have done some of these tests already. It might help us work out what area is going wrong in Godot and can be improved.

### Tests
The best tests for this kind of thing involve totally removing physics / jitter fix / smoothing from the equation. You can't begin to examine these until the fundamentals are working.

The kind of tests I would use involve moving a simple object (ball, cube) across the screen from left to right, with a wraparound so that it either rezeros the x or changes direction when it reaches the edge. You should move the ball in `_process`, and NOT in `_physics_process`.

I've just prepared a little test project that does this (there is more info in the enclosed readme):
[LawnjellyJitterTests.zip](https://github.com/godotengine/godot/files/4532634/LawnjellyJitterTests.zip)

_(Press F8 to close the window if you are running from fullscreen.)_

There are 2 important considerations here:
1) The input delta - ideally this should match the delta between starting to simulate a frame and it appearing on screen. In practice this is difficult to achieve.
2) The output delta - the actual delta between the start of simulation and the time it appears on screen.

If both of these are perfect, then you should get perfectly smooth motion (albeit with possibly irregular frame display times .. motion will probably look better with regular display times, all other things being equal).

In the test project there are 2 tests, a pink box and white box.

### White box
This test fixes the input delta, and examine the effect due to output delta. In your object script you might have something like (ignoring the wraparound):
```
_process(delta):
    position.x += 0.1
```
This is working on the assumption that with a simple scene with 1 object, the CPU time should be approximately equal for each frame. Let us know how this appears in terms of jitter.

### Pink box
This uses the input delta in the movement. This is something more like:
```
_process(delta):
    position.x += (delta * 0.1)
```
Hopefully the input deltas will be relatively constant, and you will get a result similar to the white box.

### Frame deltas
The project can also record the input deltas. If PRINT_FRAME_TIMES is set to true in Root.gd, it will log and print out input deltas. This can be helpful to diagnose whether problems are with input delta or output delta.

Be sure to test these in exported projects rather than just in the godot editor, and also in fullscreen as well as windowed. Fullscreen can work quite differently from windowed, depending on the OS.

**Comment by pip-byte** [comment]
Thank you for your detailed and extensive answer.
I testet your project on MacOS both in the editor and as an exported file.
The two boxes seem to run smooth but jittering happens regularly every second or third pass.

Frame Time output (running full screen in the editor):
https://pastebin.com/27YqA6JJ

Profiler output (from a second pass) in the editor:
<img width="1683" alt="ProfilerJitterTest" src="https://user-images.githubusercontent.com/58225516/80280798-b1c12480-8706-11ea-9f04-8965079001ad.png">

Screengrab running fullscreen in an exported project. 
https://gofile.io/?c=HNFmSC

I will try the test on iOS later.

**Comment by lawnjelly** [comment]
A lot of the timings seemed good, at bang on 60fps. Which begs the question, how are you getting 60fps on a 120fps capable monitor? Are you using `debug/settings/fps/force_fps` here?

I would advise against using `force_fps`, the reason being is it uses an incredibly inaccurate OS sleep to achieve this, which will throw everything off. It could work for battery saving at a stretch on mobile but it's practically guaranteed to screw up timing.

**Comment by lawnjelly** [comment]
Anyway the input pattern seems to be a period of solid (long) 1/60 second frames followed by some more random flicks between medium and short frames. This could easily be a result of using `force_fps` and sleeps. If not, it could be something else is happening periodically to screw up the input timings.

In the editor there is also lots of debug stuff that happens periodically (network communications etc), it could be this, so it might be a good idea to try measurements on an exported release project, you might have to run the exported project from a console to see the log though (I'm hoping you can do this on Mac, I don't know this OS).

**Comment by pip-byte** [comment]
Settings for your test project:

- force_fps = 0
- physics jitter fix = 0
- allow hidpi = true

I exported the project to iOS and ran it on my iPad. The stuttering looks similar to the results on the the iMac (which only has a 60Hz display so maybe the variable display rate is not the culprit here).
Console Log from Xcode. iPad Pro 12.9, 4th Gen (2020), iOS 13.4.1 
https://pastebin.com/8u8RBrjs

I also tried switching from GLES2 to GLES3 and the stuttering is much (!) worse (editor/fullscreen).
https://pastebin.com/n6cLQbvJ

**Comment by lawnjelly** [comment]
The timings from the ipad look fairly sensible. But maybe that section didn't capture a stutter.

The timings with GLES3 look all over the place. It looks like it's just discarding frames (which isn't something we are expecting), I don't know what the OS specific code is on the Mac, maybe that will yield some clues.

**Comment by pip-byte** [comment]
The same test works flawlessly on my old (non retina screen) Macbook Air running the same MacOS version so I don't think it is linked to the OS.

It would be great if someone could test your demo project on other devices with variable display refresh rates ..

- Android Phones like the Samsung Galaxy S20 or the Oneplus 8 Pro
- external displays

.. and see if the stuttering appears there as well.

**Comment by Calinou** [comment]
@_CHANGEME Can you (or anyone else) still reproduce this bug in [Godot 3.5](https://godotengine.org/article/godot-3-5-cant-stop-wont-stop) or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by pip-byte** [comment]
It seems to work fine since 3.4 - even on modern iOS devices with retina display.

The Godot editor is also stutter free but the frame rate tries to catch up every few seconds and the game runs at about 200% or so for a moment.

**Comment by Calinou** [comment]
> The Godot editor is also stutter free but the frame rate tries to catch up every few seconds and the game runs at about 200% or so for a moment.

Please open a separate issue for this (ideally with a video recording included).
